src/control/lib/hardware/pci_test.go

```diff
@@ -0,0 +1,534 @@
+				*MustNewPCIAddress("0000:01:03.1"): []*PCIDevice{},
+			},
+			expResult: []string{"0000:01:01.1", "0000:01:02.1", "0000:01:03.1"},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			result := tc.devices.Keys()
+			t.Logf("result: %v", result)
+			resultStr := make([]string, len(result))
+			for i, key := range result {
+				resultStr[i] = key.String()
+			}
+
+			if diff := cmp.Diff(tc.expResult, resultStr); diff != "" {
+				t.Fatalf("(-want, +got)\n%s\n", diff)
+			}
+		})
+	}
+}
+
+func TestHardware_PCIDevices_Add(t *testing.T) {
+	for name, tc := range map[string]struct {
+		devices   PCIDevices
+		newDev    *PCIDevice
+		expResult PCIDevices
+	}{
+		"nil": {},
+		"add nil Device": {
+			devices:   PCIDevices{},
+			expResult: PCIDevices{},
+		},
+		"add to empty": {
+			devices: PCIDevices{},
+			newDev: &PCIDevice{
+				Name:    "test1",
+				Type:    DeviceTypeNetInterface,
+				PCIAddr: *MustNewPCIAddress("0000:01:01.01"),
+			},
+			expResult: PCIDevices{
+				*MustNewPCIAddress("0000:01:01.01"): {
+					{
+						Name:    "test1",
+						Type:    DeviceTypeNetInterface,
+						PCIAddr: *MustNewPCIAddress("0000:01:01.01"),
+					},
+				},
+			},
+		},
+		"add to existing": {
+			devices: PCIDevices{
+				*MustNewPCIAddress("0000:01:01.01"): {
+					{
+						Name:    "test1",
+						Type:    DeviceTypeNetInterface,
+						PCIAddr: *MustNewPCIAddress("0000:01:01.01"),
+					},
+				},
+				*MustNewPCIAddress("0000:01:02.01"): {
+					{
+						Name:    "test2",
+						Type:    DeviceTypeUnknown,
+						PCIAddr: *MustNewPCIAddress("0000:01:02.01"),
+					},
+				},
+			},
+			newDev: &PCIDevice{
+				Name:    "test3",
+				Type:    DeviceTypeOFIDomain,
+				PCIAddr: *MustNewPCIAddress("0000:01:01.01"),
+			},
+			expResult: PCIDevices{
+				*MustNewPCIAddress("0000:01:01.01"): {
+					{
+						Name:    "test1",
+						Type:    DeviceTypeNetInterface,
+						PCIAddr: *MustNewPCIAddress("0000:01:01.01"),
+					},
+					{
+						Name:    "test3",
+						Type:    DeviceTypeOFIDomain,
+						PCIAddr: *MustNewPCIAddress("0000:01:01.01"),
+					},
+				},
+				*MustNewPCIAddress("0000:01:02.01"): {
+					{
+						Name:    "test2",
+						Type:    DeviceTypeUnknown,
+						PCIAddr: *MustNewPCIAddress("0000:01:02.01"),
+					},
+				},
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			tc.devices.Add(tc.newDev)
+
+			if diff := cmp.Diff(tc.expResult, tc.devices); diff != "" {
+				t.Fatalf("(-want, +got)\n%s\n", diff)
+			}
+		})
+	}
+}
+
+func TestHardware_PCIBus(t *testing.T) {
+	isNilErr := errors.New("is nil")
+
+	for name, tc := range map[string]struct {
+		bus    *PCIBus
+		dev    *PCIDevice
+		op     string
+		expErr error
+	}{
+		"add device to nil bus": {
+			dev:    mockPCIDevice("test"),
+			op:     "add",
+			expErr: isNilErr,
+		},
+		"add nil device": {
+			bus:    mockPCIBus(),
+			op:     "add",
+			expErr: isNilErr,
+		},
+		"nil bus contains device": {
+			dev: mockPCIDevice("test"),
+			op:  "nil contains",
+		},
```

Comment:
Personally would prefer to break Contains tests out into their own test function.

Reply:
Yeah, I got lazy. I'll do that in my next pass.